### PR TITLE
Fix title selection for untiled files

### PR DESCRIPTION
**Pull request by ArkadiuszMichalski**
Fix https://github.com/notepad-plus-plus/notepad-plus-plus/issues/13194, https://github.com/notepad-plus-plus/notepad-plus-plus/issues/8677 and https://github.com/notepad-plus-plus/notepad-plus-plus/issues/11704.

### CODE REVIEW

**Comment by Yaron10** [comment]
@ArkadiuszMichalski,

Thank you for your work. 👍 

After a *quick test* - your PR fixes the issues and doesn't seem to cause any regression.
But this kind of change requires more than a "quick test" and a better understanding of the code (i.e. not myself).

Still, allow me to ask: if `toMainOrSubEditView == false` - should `nextUntitledNewNumber()` be called at all?

***

Related:
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/11704.

**Comment by ArkadiuszMichalski** [comment]
@Yaron10 
>Still, allow me to ask: if toMainOrSubEditView == false - should nextUntitledNewNumber() be called at all?

From what I can see every buffer has `_fileName` set at creation time. Some are created and invisibly but for them we also need to give a name. To make it easier we use `nextUntitledNewNumber()` so the name is incremented (`New 1`, `New 2`) instead of something fixed (like `New` for all of them). Thanks to this we do not have to take care of it ourselves. Buffers not loaded into `MainEditView` or `SubEditView` (i.e. loaded into other View or not at all) also have some value in `_fileName`. Automatic generation of this value makes it easier.

https://github.com/notepad-plus-plus/notepad-plus-plus/search?q=nextUntitledNewNumber&type=code

**Comment by Yaron10** [comment]
@ArkadiuszMichalski,

Thank you for the explanation and also for fixing https://github.com/notepad-plus-plus/notepad-plus-plus/issues/11704. 👍 

I'd like to emphasize again that I don't fully understand the relevant code.
But I still think that my question deserves some further investigation. :)

https://github.com/notepad-plus-plus/notepad-plus-plus/blob/99e7c5a84cc994b795b7b5e7acfa4943b9b03ed4/PowerEditor/src/ScintillaComponent/Buffer.cpp#L1335

Should `nextUntitledNewNumber()` be called for the invisible view or the Search Results?

Apart from that - just replacing
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/99e7c5a84cc994b795b7b5e7acfa4943b9b03ed4/PowerEditor/src/ScintillaComponent/Buffer.cpp#L1267

with

```
if (buf->isUntitled() && (buf->_referees[0])->isVisible())
		{
			if (buf->indexOfReference(_pNotepadPlus->_pEditView) > -1 || buf->indexOfReference(_pNotepadPlus->_pNonEditView) > -1)
			{
				generic_string newTitle = ((NppParameters::getInstance()).getNativeLangSpeaker())->getLocalizedStrFromID("tab-untitled-string", UNTITLED_STR);
				TCHAR *numberStr = buf->_fileName + newTitle.length();
				int usedNumber = _wtoi(numberStr);
				usedNumbers.push_back(usedNumber);
			}
		}
	}
```

seems to solve https://github.com/notepad-plus-plus/notepad-plus-plus/issues/8677 (i.e. without any further changes).

Let's wait for Don's review.

File: PowerEditor/src/ScintillaComponent/Buffer.cpp [hunk 1]
```diff
@@ -1258,16 +1258,25 @@ SavingStatus FileManager::saveBuffer(BufferID id, const TCHAR * filename, bool i
...
+			for (size_t k = 0; k < buf->_referees.size(); k++)
+			{
+				if (buf->_referees[k]->isVisible())
+				{
+					isVisible = true;
+					break;
+				}
+			}
 			// if untitled document is invisible, then don't put its number into array (so its number is available to be used)
-			if ((buf->_referees[0])->isVisible())
```
Comment (donho) [inline PowerEditor/src/ScintillaComponent/Buffer.cpp:1270]: Please use the following code:
```cpp
		if (buf->isUntitled())
		{
			bool isVisible = false;
			for (size_t k = 0; k < buf->_referees.size(); k++)
			{
				if (buf->_referees[k]->isVisible())
				{
					isVisible = true;
					break;
				}
			}

			if (isVisible)
			{
				if (buf->indexOfReference(_pNotepadPlus->_pEditView) > -1 || buf->indexOfReference(_pNotepadPlus->_pNonEditView) > -1)
				{
					generic_string newTitle = ((NppParameters::getInstance()).getNativeLangSpeaker())->getLocalizedStrFromID("tab-untitled-string", UNTITLED_STR);
					TCHAR* numberStr = buf->_fileName + newTitle.length();
					int usedNumber = _wtoi(numberStr);
					usedNumbers.push_back(usedNumber);
				}
			}
		}
```
to replace the old code:
```cpp
		if (buf->isUntitled())
		{
			// if untitled document is invisible, then don't put its number into array (so its number is available to be used)
			if ((buf->_referees[0])->isVisible())
			{
				generic_string newTitle = ((NppParameters::getInstance()).getNativeLangSpeaker())->getLocalizedStrFromID("tab-untitled-string", UNTITLED_STR);
				TCHAR *numberStr = buf->_fileName + newTitle.length();
				int usedNumber = _wtoi(numberStr);
				usedNumbers.push_back(usedNumber);
			}
		}
```

The other modification is not necessary.

**Review by donho — CHANGES_REQUESTED** [review]
@Yaron10 's simpler solution fixes #13194 and #8677. However it doesn't fix #11704.

I have to combine @ArkadiuszMichalski 's code to make #8677 fixed:
```
			bool isVisible = false;
			for (size_t k = 0; k < buf->_referees.size(); k++)
			{
				if (buf->_referees[k]->isVisible())
				{
					isVisible = true;
					break;
				}
			}
```

Thank you for the contribution from both of you :)

**Comment by Yaron10** [comment]
@ArkadiuszMichalski,

So it seems your approach of modifying `nextUntitledNewNumber()` was correct. :)
Thanks again for your work. 👍 

@donho,

Thank you for reviewing and optimizing. 👍

**Review by donho — APPROVED** [review]

**Pull request closed without merging**